<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_1 = f'''select *
            from vacancies
         '''
# чтобы посчитать количество вакансий, нужно конечно сделать запрос с count(*),
# но в начале хочется посмотреть на вид и содержание таблиц, поэтому выбиралю все строки
# количество строк равно количеству вакансий

In [5]:
# результат запроса
vacancies_df = pd.read_sql_query(query_1, connection)
vacancies_df

id                                               name  \
0      55312386                                Компьютерный Мастер   
1      55843014                            Системный администратор   
2      54525964                      Lead Java Developer to Poland   
3      54525965                      Lead Java Developer to Poland   
4      55354053  Специалист службы поддержки с техническими зна...   
...         ...                                                ...   
49192  55589514           Специалист по информационным технологиям   
49193  55029668                     Руководитель процессного офиса   
49194  66172263                         Начальник сервисной службы   
49195  66180086                             Senior IOS разработчик   
49196  55547842                                   DataOps Engineer   

                                              key_skills          schedule  \
0      Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1      Средства криптографической защиты информации\t...       Полный день   
2        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3        Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                                   None  Удаленная работа   
...                                                  ...               ...   
49192  Электронный документооборот\t1С: Предприятие 8...       Полный день   
49193  Оптимизация бизнес-процессов\tСтратегическое м...       Полный день   
49194  Организаторские навыки\tУправление персоналом\...       Полный день   
49195  Swift\tRealm\tiOS\tАнглийский язык\tMVP\tMVVM\...       Полный день   
49196                                               None  Удаленная работа   

               experience           employment  salary_from  salary_to  \
0               Нет опыта     Полная занятость      64000.0        NaN   
1           От 3 до 6 лет     Полная занятость          NaN        NaN   
2           От 3 до 6 лет     Полная занятость          NaN        NaN   
3           От 3 до 6 лет     Полная занятость          NaN        NaN   
4               Нет опыта  Частичная занятость      15000.0        NaN   
...                   ...                  ...          ...        ...   
49192       От 3 до 6 лет     Полная занятость          NaN        NaN   
49193       От 3 до 6 лет     Полная занятость          NaN        NaN   
49194  От 1 года до 3 лет     Полная занятость      60000.0        NaN   
49195  От 1 года до 3 лет     Полная занятость     146456.0   244094.0   
49196       От 3 до 6 лет     Полная занятость          NaN        NaN   

       area_id  employer_id  
0         1516      5724811  
1          130      4903713  
2          160        69961  
3          159        69961  
4         1955         1740  
...        ...          ...  
49192       66      5109332  
49193       26        51333  
49194     1930      4094700  
49195      160      4469939  
49196      160      4978627  

[49197 rows x 10 columns]

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_2 = f'''select * 
            from employers
         '''
# Аналогично, смотрим всю таблицу, количество работодателей равно количеству строк талбицы

In [7]:
# результат запроса
employers_df = pd.read_sql_query(query_2, connection)
employers_df

id                                               name  area
0         2393                   Программный Продукт, ИТ-компания     1
1        72977                                          БАРС Груп    88
2         3155  Бест, Торгово-производственная компания, Екате...     3
3          675                                   КОРУС Консалтинг     2
4      1840010  филиал ФКУ Налог-Сервис ФНС России в Республик...    88
...        ...                                                ...   ...
23496  1441774                                              МарКо    88
23497  4718575                                      MKomov Studio    73
23498  3702770                                  АКИБ Ипотека-банк  2759
23499  5636092                                        ГИМ-КЛИНИКА    99
23500  2064382                                             Рыббла   104

[23501 rows x 3 columns]

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3 = f'''select *
             from areas
             '''
# количество регионов равно количеству строк

In [9]:
# результат запроса
areas_df=pd.read_sql_query(query_3, connection)
areas_df

id              name
0     2758           Тбилиси
1        8            Майкоп
2     1180          Нерюнгри
3     1240       Новокузнецк
4        2   Санкт-Петербург
...    ...               ...
1357  5043  Звёздный Городок
1358  3934       Воротынск\n
1359  4825             Чемал
1360  3708         Усть-Нера
1361  2432         Каневская

[1362 rows x 2 columns]

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_4 = f'''select *
           from industries
           '''
# количество сфер деятельности равно количеству строк

In [11]:
# результат запроса
industries_df = pd.read_sql_query(query_4, connection)
industries_df

id                                               name
0      7.540                Разработка программного обеспечения
1      7.539  Системная интеграция,  автоматизации технологи...
2     27.550              Безалкогольные напитки (производство)
3     27.551  Безалкогольные напитки (продвижение, оптовая т...
4     13.664             Управление и эксплуатация недвижимости
..       ...                                                ...
289   37.626             Ассоциация в сфере культуры, искусства
290   42.603                              Табак  (производство)
291   29.532                           Рыболовство, рыбоводство
292   37.627                      Благотворительная организация
293  388.505  Оборудование для деревообработки, заготовки ле...

[294 rows x 2 columns]

In [16]:
# Сделаем дополнительный запрос о работадателях, у которых нет размещенных вакансий
query_e = f'''select count(e.id) 
            from employers e
            left join vacancies v on e.id=v.employer_id
            where v.id is NULL 
         '''

In [18]:
emp_count_null = pd.read_sql_query(query_e, connection)
print('У {} работодателей нет размещенных вакансий'.format(emp_count_null['count'][0]))

У 8595 работодателей нет размещенных вакансий


***

## выводы по предварительному анализу данных

По запросам видим, что на 23,5 тысячи работодателей около 50 тысяч вакансий - не так уж много, поэтому сделаем доп. запрос и посмотрим сколько работодателей не имеют активных вакансий.

Из доп. запроса видим, что 8,5 тысяч (около трети) работодателей из базы не размещают вакансии. Это может говорить о том, что либо многие закрылись, либо размещают вакансии под другим юр. лицом, в предприятиях минимальная текучка кадров и нет необходимости в новых сотрудниках. Так же может быть признаком снижения числа вакансий, сокращения рабочих мест. (Здесь интересно было бы посмотреть данные за более поздние периоды)

По запросу видим, что около 300 сфер деятельности, достаточно разнообразно.
Около 1,3 тысячи регионов указаны, что очень много, то есть вакансии представлены не только крупными городами, но и очень много небольших населенных пунктов, так же в регионах видим страны города ближнего зарубежья.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [19]:
# текст запроса
query_5 = f'''select a.name area, count(v.id) cnt
           from vacancies v 
           join areas a on v.area_id=a.id
           group by area
           order by cnt desc
           '''

In [23]:
# результат запроса
area_vac_df=pd.read_sql_query(query_5, connection)
area_vac_df.head(6)

area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892
5     Екатеринбург  1698

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [6]:
# текст запроса
query_6 = f'''select count(*)
           from vacancies
           where salary_from is not NULL or salary_to is not NULL
           '''

In [17]:
# результат запроса
salary_fill=pd.read_sql_query(query_6, connection)
print('У {} вакансий заполнены поля с зарплатой'.format(salary_fill['count'][0]))

У 24073 вакансий заполнены поля с зарплатой


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [24]:
# текст запроса
query_7 = f'''select round(avg(salary_from),0) avg_from, round(avg(salary_to),0) avg_to
           from vacancies
           '''

In [25]:
# результат запроса
salary_avg=pd.read_sql_query(query_7, connection)
salary_avg

avg_from    avg_to
0   71065.0  110537.0

In [75]:
# Смотря на предыдуший запрос, хочется посмотреть на среднюю предлагаемую зарплату по отрослям.
# Поэтому сделаем дополнительный запрос: выведем название отрасли, среднюю зарплату (посмотрим на нижнюю границу) и
# количество вакансии в отрасли, уберем отрасли, где вакансий меньше 10
# отсортируем по средней ЗП в порядке убивания

query_i = f'''select i.name industry, round(avg(v.salary_from),0) avg_from, count(v.id)
           from vacancies v
           join employers_industries ei on v.employer_id=ei.employer_id
           join industries i on ei.industry_id=i.id
           where (v.salary_from is not NULL)
           group by i.name
           having count(v.id)>=10
           order by 2 desc
                     '''

In [76]:
# результат запроса
salary_avg_all=pd.read_sql_query(query_i, connection)
salary_avg_all.head(20)

industry  avg_from  count
0           Агрохимия (продвижение, оптовая торговля)  403077.0     13
1                            Агрохимия (производство)  338625.0     16
2                                      Игорный бизнес  147993.0     11
3                                          Зоомагазин  137419.0     31
4                   Корма для животных (производство)  136024.0     29
5   Бытовая химия, парфюмерия, косметика  (произво...  133807.0     70
6                                  Кадровые агентства  118131.0    575
7                      Учреждение соц.помощи и защиты  116786.0     14
8                           Ветеринарная деятельность  113171.0     13
9       Лакокрасочная продукция, сырье (производство)  106190.0     21
10  Управляющая, инвестиционная компания (управлен...  105287.0     71
11  Финансово-кредитное посредничество (биржа, бро...  100577.0    166
12  Бытовая химия, парфюмерия, косметика  (продвиж...  100576.0    131
13                               Снеки (производство)   98537.0     20
14  Рыба и морепродукты (продвижение, оптовая торг...   97813.0     16
15  Корма для животных (продвижение, оптовая торго...   97447.0     53
16  Лакокрасочная продукция, сырье (продвижение, о...   96662.0     27
17  Инфраструктурная компания в энергетике (СО ЕЭС...   95749.0    124
18                          Железнодорожные перевозки   93176.0     41
19           Консалтинг в области энергоэффективности   92218.0     15

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [26]:
# текст запроса
query_8 = f'''select schedule, employment, count(*) 
           from vacancies
           group by 1,2
           order by 3 desc
           '''

In [9]:
# результат запроса
sch_empl=pd.read_sql_query(query_8, connection)
sch_empl

schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [18]:
# текст запроса
query_9 = f'''select experience
           from vacancies
           where experience is not NULL
           group by 1
           order by count(experience)
           '''

In [19]:
# результат запроса
exp=pd.read_sql_query(query_9, connection)
exp

experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

***

## выводы по детальному анализу вакансий

Из запроса видим, что меньше всего вакансий с опытом работы более 6 лет. С чем это может быть связано? Можно предположить, что специалистов с большим опытом не так часто ищут с помощь hh, чаще ищут в специализированных сообществах (чатах). В целом,  хорошего опытного специалиста компания будет стараться сохранить у себя, поэтому, если сотрудник работает от 3 до 6 лет в компании и не уходит, значит его все устраивает, соответсвенно запросов на вакансии с таким опытом меншье. 
Больше всего вакансий для категории с опытом от 1 до 3 лет, что логично. Сотрудник уже более-менее опытный, требудет меньше времени на обучение и адаптацию в компании, но менее требовательный, чем соискатель с опытом от 3 лет и более.

Из запроса о количестве вакансий в зависимости от сочетания типа графика и типа трудоустройства видим, что наиболее популярное сочетание - Полный день/Полная занятость, что логично. На втором месте - Удаленная работа/Полная занятость, можно сделать вывод, что все большое количество вакансий с таким типом трудоустройства и скорее всего они будут увеличиваться. Ведь удаленная работа это экнономия работодателя на аренду офиса, оборудования и каких-то схожих затратах. (Здесь также интересно посмотреть данные за прошлые периоды).

Из доп. запроса видим, что наиболее выские зарплаты предлагают на производстве, в отповой торговле. По количеству ванасий с высокой ЗП, больше всего у кадровых агентств, их можно исключить, ведь они предлагают вакансий в разных сферах.

Из запроса доп. запроса видим, что у половины вакансий не заполнены поля с зарплатой. Часто поле ЗП не заполняют, если на выставленную ЗП нет или очень мало откликов и/или работодатель хочет посмотреть на какую ЗП согласятся соискатели, чтоб предложить оптимальную ЗП и найти подходящего работника (это может быть для новых позиций в компании). 

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [77]:
# текст запроса
query_10 = f'''(select e.name
           from vacancies v
           join employers e on e.id=v.employer_id
           group by 1
           order by count(v.id) desc
            limit 1)
           union all
           (select e.name
           from vacancies v
           join employers e on e.id=v.employer_id
           group by 1
           order by count(v.id) desc
           offset 4
           limit 1)
           '''

In [78]:
# результат запроса
empl_vac=pd.read_sql_query(query_10, connection)
empl_vac

name
0         Яндекс
1  Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [80]:
# текст запроса
query_11 = f'''select a.name city, count(distinct e.id) count_employers, count(distinct v.id) count_vacancies
           from areas a
           left join employers e on a.id=e.area
           left join vacancies v on a.id=v.area_id
           group by city 
           having count(distinct v.id)=0
           order by 2 desc
           '''

In [81]:
# результат запроса
empl_vac=pd.read_sql_query(query_11, connection)
empl_vac

city  count_employers  count_vacancies
0                            Россия              410                0
1                         Казахстан              207                0
2                Московская область               75                0
3                Краснодарский край               19                0
4                          Беларусь               18                0
..                              ...              ...              ...
588                        Кувандык                0                0
589                        Куйбышев                0                0
590  Куйбышево (Ростовская область)                0                0
591                        Култаево                0                0
592                        Кульсары                0                0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [82]:
# текст запроса
query_12 = f'''select e.name employers, count(distinct v.area_id) count_areas
           from employers e
           join vacancies v on v.employer_id=e.id
           group by employers
           order by count_areas desc
           '''

In [83]:
# результат запроса
empl_areas=pd.read_sql_query(query_12, connection)
empl_areas

employers  count_areas
0                      Яндекс          181
1                  Ростелеком          152
2                  Спецремонт          116
3      Поляков Денис Иванович           88
4                   ООО ЕФИН            71
...                       ...          ...
14761                  UniSol            1
14762            UNISTORY LLC            1
14763                   UNIT6            1
14764     United Distribution            1
14765              UNITEL LLC            1

[14766 rows x 2 columns]

In [23]:
# Сделаем доп. запрос чтоб посмотреть, что за вакансии размещает работодатель, находящийся на месте по количеству
query_t = f'''select * 
           from employers e
           join vacancies v on v.employer_id=e.id
           where e.name='Поляков Денис Иванович'
          '''

In [24]:
q=pd.read_sql_query(query_t, connection)
q

id                    name  area        id  \
0    5130287  Поляков Денис Иванович    22  55402339   
1    5130287  Поляков Денис Иванович    22  53316670   
2    5130287  Поляков Денис Иванович    22  54290525   
3    5130287  Поляков Денис Иванович    22  54303686   
4    5130287  Поляков Денис Иванович    22  66095025   
..       ...                     ...   ...       ...   
122  5130287  Поляков Денис Иванович    22  55186762   
123  5130287  Поляков Денис Иванович    22  55836122   
124  5130287  Поляков Денис Иванович    22  54534859   
125  5130287  Поляков Денис Иванович    22  55304208   
126  5130287  Поляков Денис Иванович    22  54427430   

                             name key_skills     schedule          experience  \
0             Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
1             Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
2             Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
3             Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
4             Компьютерный мастер       None  Полный день           Нет опыта   
..                            ...        ...          ...                 ...   
122           Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
123           Компьютерный мастер       None  Полный день           Нет опыта   
124           Компьютерный мастер       None  Полный день  От 1 года до 3 лет   
125  Ученик компьютерного мастера       None  Полный день           Нет опыта   
126           Компьютерный мастер       None  Полный день  От 1 года до 3 лет   

           employment  salary_from  salary_to  area_id  employer_id  
0    Полная занятость        60000        NaN     1141      5130287  
1    Полная занятость        60000        NaN       47      5130287  
2    Полная занятость        60000        NaN       56      5130287  
3    Полная занятость        60000        NaN       70      5130287  
4    Полная занятость        40000    55000.0       20      5130287  
..                ...          ...        ...      ...          ...  
122  Полная занятость        60000        NaN     1243      5130287  
123  Полная занятость        50000        NaN       58      5130287  
124  Полная занятость        60000        NaN      212      5130287  
125        Стажировка        30000    50000.0        4      5130287  
126  Полная занятость        60000        NaN       25      5130287  

[127 rows x 13 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [84]:
# текст запроса
query_13 = f'''select count(*)
           from employers e
           left join employers_industries ei on e.id=ei.employer_id
           where industry_id is NULL
           '''

In [85]:
# результат запроса
empl_count=pd.read_sql_query(query_13, connection)
empl_count
print('У {} работодателей не указана сфера деятельности'.format(empl_count['count'][0]))

У 8419 работодателей не указана сфера деятельности


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [86]:
# текст запроса
query_14 = f'''select e.name employers
           from employers e
           left join employers_industries ei on e.id=ei.employer_id
           group by employers
           having count(ei.employer_id)=4
           order by employers
           offset 2
           limit 1
           '''

In [87]:
# результат запроса
empl_count2=pd.read_sql_query(query_14, connection)
print('{} находится на третьем месте среди работодателей, у которых указано 4 сферы деятельности'.format(empl_count2['employers'][0]))

2ГИС находится на третьем месте среди работодателей, у которых указано 4 сферы деятельности


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [88]:
# текст запроса
query_15 = f'''select count(e.name) cnt
           from employers e
           join employers_industries ei on e.id=ei.employer_id
           join industries i on i.id=ei.industry_id
           where i.name='Разработка программного обеспечения'
           '''

In [89]:
# результат запроса
empl_PO=pd.read_sql_query(query_15, connection)
print('У {} работодателей в качестве сферы деятельности указано -"Разработка программного обеспечения"'.format(empl_PO['cnt'][0]))

У 3553 работодателей в качестве сферы деятельности указано -"Разработка программного обеспечения"


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [4]:
# код для получения списка городов-милионников
!pip install beautifulsoup4 
from bs4 import BeautifulSoup

In [5]:
import requests
from bs4 import BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем адрес страницы
response = requests.get(url) 
page = BeautifulSoup(response.text, 'html.parser')
lst = page.find_all('a')
lst_cities=tuple([link.text for link in lst[46:63]])
lst_cities

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 '[a]',
 'Пермь',
 'Волгоград')

In [76]:
# текст запроса
query_16 = f'''(select a.name, count(v.id) cnt
           from vacancies v 
           join areas a on a.id=v.area_id
           join employers e on e.id=v.employer_id
           where e.name='Яндекс'
           group by a.name
           having a.name in ('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 '[a]',
 'Пермь',
 'Волгоград'))
 union all
 (select 'Total', count(*)
 from vacancies v 
           join areas a on a.id=v.area_id
           join employers e on e.id=v.employer_id
           where e.name='Яндекс' and a.name in ('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Челябинск',
 'Красноярск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Омск',
 'Краснодар',
 'Воронеж',
 '[a]',
 'Пермь',
 'Волгоград'))
  order by 2
           '''

In [79]:
# результат запроса
ya=pd.read_sql_query(query_16, connection)
ya

name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

## выводы по анализу работодателей
Мы видим, что нет вакансий, если регион работодателя указан в общем - страна, область, и также нет вакансий в маленький городах. Возможно такие регионы не стоит указывать, лучше конкретный город. Если соискатель ищет вакансию, вряд ли будет искать по региону Россия или Московска область, скорее всего будет искать конкретный город.

Большое колисество работодателя (около трети) не указывают сферу деятельности в вакансии, что затрудняет поиск для соискателей, так как многие ищут вакансии, выбирая конкретные сферы деятельности. При размещении вакансии стоит рекомендовать работодателям, чтоб указывали сферу деятельности, что увеличит число просмотров вакансии.

Выполнила доп. запрос, чтоб проверить работодателя 'Поляков Денис Иванович' (это обычно ИП), при этом размещает вакансии в 88 регионах, что очень много для ИП (выглядит подозрительно). Из запроса видим, что это одни и те же вакансии 'компьютерный мастер'. Вполне может быть, что такая сеть мастерских.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [90]:
# текст запроса
query_17 = f'''select count(*)
           from vacancies
           where lower(name) like '%data%' or lower(name) like '%данн%'
           '''

In [91]:
# результат запроса
cnt_data_vac=pd.read_sql_query(query_17, connection)
print('{} ваканисий(я) имеет отношение к данным'.format(cnt_data_vac['count'][0]))

1771 ваканисий(я) имеет отношение к данным


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
 * 'data scientist'
 * 'data science'
 * 'исследователь данных'
 * 'ML' (здесь не нужно брать вакансии по HTML)
 * 'machine learning'
 * 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [92]:
# текст запроса
query_18 = f'''select count(name)
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%'
           or lower(name) like '%машинн%обучен%'
           or 
           (name like '%ML%' and lower(name) not like '%html%'))
           and 
           (employment='Стажировка' or experience='Нет опыта' or lower(name) like '%junior%')
           '''

In [94]:
# результат запроса
cnt_ds=pd.read_sql_query(query_18, connection)
print('{} - количество вакансий, подходящих для начинающего дата-сайентиста'.format(cnt_ds['count'][0]))

51 - количество вакансий, подходящих для начинающего дата-сайентиста


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [95]:
# текст запроса
query_19 = f'''select count(*)
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%'
           or lower(name) like '%машинн%обучен%'
           or (name like '%ML%'  and lower(name) not like '%html%'))
           and (lower(key_skills) like '%sql%' or lower(key_skills) like '%postgres%')
           '''

In [96]:
# результат запроса
cnt_sk=pd.read_sql_query(query_19, connection)
print('{} - количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres'.format(cnt_sk['count'][0]))

201 - количество вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres


In [10]:
# сделаем доп. запрос и посмотрим сколько всего вакансий, в которых указан SQL в качестве ключевого навыка
query_sql = f'''select count(*)
           from vacancies
           where lower(key_skills) like '%sql%' or lower(key_skills) like '%postgres%'
           '''

In [11]:
cnt_sql=pd.read_sql_query(query_sql, connection)
print('{} - количество вакансий, в которых в качестве ключевого навыка указан SQL или postgres'.format(cnt_sql['count'][0]))

10888 - количество вакансий, в которых в качестве ключевого навыка указан SQL или postgres


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [97]:
# текст запроса
query_20 = f'''select count(*)
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%'
           or lower(name) like '%машинн%обучен%'
           or 
           (name like '%ML%'  and lower(name) not like '%html%'))
           and 
           (lower(key_skills) like '%python%')
           '''

In [98]:
# результат запроса
cnt_pth=pd.read_sql_query(query_20, connection)
print('{} - количество вакансий для DS, в которых в качестве ключевого навыка указан Python'.format(cnt_pth['count'][0]))

351 - количество вакансий для DS, в которых в качестве ключевого навыка указан Python


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [99]:
# текст запроса
query_21 = f'''select avg(length(key_skills) - length(replace(key_skills,'\t',''))+1)
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%'
           or lower(name) like '%машинн%обучен%'
           or (name like '%ML%'  and lower(name) not like '%html%'))
           '''

In [100]:
# результат запроса
cnt_skills=pd.read_sql_query(query_21, connection)
cnt_skills

avg
0  6.406032

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [101]:
# текст запроса
query_22 = f'''select experience, round(avg(COALESCE((salary_from+salary_to)/2, salary_from, salary_to)),0) avg_sal
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%' or lower(name) like '%машинн%обучен%'
           or (name like '%ML%'  and lower(name) not like '%html%'))
           and (salary_from is not NULL or salary_to is not NULL)
           group by experience
           '''

In [102]:
# результат запроса
avg_sel=pd.read_sql_query(query_22, connection)
avg_sel

experience   avg_sal
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

In [104]:
# Сделаем доп. запрос, чтоб посмотреть средние ЗП по всем вакансиям
query_s = f'''select experience, round(avg(COALESCE((salary_from+salary_to)/2, salary_from, salary_to)),0) avg_sal
           from vacancies
           group by experience
           '''

In [105]:
avg_sel_all=pd.read_sql_query(query_s, connection)
avg_sel_all

experience   avg_sal
0         Более 6 лет  166256.0
1       От 3 до 6 лет  133211.0
2           Нет опыта   40472.0
3  От 1 года до 3 лет   76541.0

In [106]:
# Сделаем доп. запрос, что посмотреть соклько всего вакансий в DS
query_c = f'''select count(*)
           from vacancies
           where (lower(name) like '%data scientist%'
           or lower(name) like '%data science%'
           or lower(name) like '%исследователь данных%'
           or lower(name) like '%machine learning%'
           or lower(name) like '%машинн%обучен%'
           or (name like '%ML%'  and lower(name) not like '%html%'))
           '''

In [107]:
count_vac_ds=pd.read_sql_query(query_c, connection)
count_vac_ds
print('{} - количество вакансий для DS'.format(count_vac_ds['count'][0]))

480 - количество вакансий для DS


***

## выводы по предметному анализу
Обратим внимание в запросе о средней ЗП в зависимости от опыта - нет вакансий с опытом более 6 лет. Сфера Data Science достаточно новая, особенно для России, поэтому люди с опытом более 6 лет это редкость и искать их вряд ли будут на hh.

Из запроса о средней ЗП для DS вакансий в зависимости от опыта и доп. запроса о среднй ЗП по всем вакансиям видим, что в DS зарплаты почти в 2 раза выше, чем средние по всем вакансиям.

Из запроса видим, что вакансий для начинающего датасайентиста немного - 51 из 480 вакансий (из доп. запроса), относящихся к DS. Делаем вывод, найти работу без опыта в этой облати довольно сложно.

В 201 из 480 вакансий для DS в качестве ключевого навыка указан SQL, что кажется мало, ведь в DS этот язык запросов важен. Поэтому выполнили доп. запрос, чтоб посмотреть в скольких вакансиях указан SQL - 10888 - очень много. Делаем вывод о важности знания этого языка запросов. Многие из этих вакансий относятся к программированию.

    P.S. Отмечу, что запросы выполнялись исходя из критериев задания. Но в этот запрос (касающийся вакансий, относящихся к DS) наверняки могли попасть лишние вакансии (критерий 'ML'), так же многие, относящиеся к области 'DS' не попадают в эту сферу (например, аналитиков данных можно отнести сюда и т.п.). 

# Общий вывод по проекту

## подведем итог исследования, обобщите выводы
#### здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

*Дополнительные запросы проводились ранее в проекте.*

Стоит рекомендовать работодателям указывать сферу деятельности, указывать конкретный город для увеличения просмотров вакансии.
Многие соискатели при поиске вакансий фильтруют по ЗП (от...), а в половине вакансий ЗП не указана, возможно некоторым работодателям все таки стоит указывать зарплатную вилку. 

Прогнозы: вакансии с удаленной работой будут увеличиваться, знание языков программирования будет более востребованы.
Что касается тенденций, то хорошо бы было сравнить с предыдущими периодами (оценить количество вакансий и ЗП за более старые периоды).

Нам стоит подумать, как улучить критерии для лучшего подсчета вакансий, относящихся к DS.   

Можно провести исследования по ключевым навыкам, вывести их отдельно и посмотреть более востребованные.